ICE: fn_ctxt: index out of bounds

<!--
[31mICE[0m: Rustc ./a.rs '' 'thread 'rustc' panicked at compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:2289:46: 'index out of bounds: the len is 1 but the index is 1'', 'thread 'rustc' panicked at compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:2289:46: 'index out of bounds: the len is 1 but the index is 1''
File: /tmp/im/a.rs
-->

auto-reduced (treereduce-rust):
````rust
trait A<T> {}

fn f<T, U, V: A<T>>(i: V, arena: T, k: U) -> (T, U) {
    f.call_once(j, k)
}


````

original:
````rust
trait A<T> {
    fn g<U>(&self, x: T, y: U) -> (T, U) { (x, y) }
}

impl From<I> for i32 { }
impl<EncodedValue> A<T> for u32 { }

fn f<T, U, V: A<T>>(i: V, arena: T, k: U) -> (T, U) {
    f.call_once(j, k)
}

pub fn main () {
    assert_eq!(f(0, 1, 2), (1, 2));
    assert_eq!(f(0, 1, 2), (1, 2));
}

````

Version information
````
rustc 1.82.0-nightly (2048386fe 2024-08-08)
binary: rustc
commit-hash: 2048386fe2898febe7315c0feb915458e41c7aa5
commit-date: 2024-08-08
host: x86_64-unknown-linux-gnu
release: 1.82.0-nightly
LLVM version: 19.1.0

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0425]: cannot find value `j` in this scope
 --> /tmp/icemaker_global_tempdir.GCP2eBzpwgpY/rustc_testrunner_tmpdir_reporting.asPRKVXEGOYI/mvce.rs:4:17
  |
4 |     f.call_once(j, k)
  |                 ^ not found in this scope

error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.GCP2eBzpwgpY/rustc_testrunner_tmpdir_reporting.asPRKVXEGOYI/mvce.rs:5:2
  |
5 | }
  |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.GCP2eBzpwgpY/rustc_testrunner_tmpdir_reporting.asPRKVXEGOYI/mvce.rs`

error[E0658]: use of unstable library feature 'fn_traits'
 --> /tmp/icemaker_global_tempdir.GCP2eBzpwgpY/rustc_testrunner_tmpdir_reporting.asPRKVXEGOYI/mvce.rs:4:7
  |
4 |     f.call_once(j, k)
  |       ^^^^^^^^^
  |
  = note: see issue #29625 <https://github.com/rust-lang/rust/issues/29625> for more information
  = help: add `#![feature(fn_traits)]` to the crate attributes to enable
  = note: this compiler was built on 2024-08-08; consider upgrading it if it is out of date

thread 'rustc' panicked at compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:2289:46:
index out of bounds: the len is 1 but the index is 1
stack backtrace:
   0:     0x7e97171eab0d - std::backtrace_rs::backtrace::libunwind::trace::h288a87c5e2679186
                               at /rustc/2048386fe2898febe7315c0feb915458e41c7aa5/library/std/src/../../backtrace/src/backtrace/libunwind.rs:116:5
   1:     0x7e97171eab0d - std::backtrace_rs::backtrace::trace_unsynchronized::h5284ad94e8c74688
                               at /rustc/2048386fe2898febe7315c0feb915458e41c7aa5/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7e97171eab0d - std::sys::backtrace::_print_fmt::h371adc3f02dab1dd
                               at /rustc/2048386fe2898febe7315c0feb915458e41c7aa5/library/std/src/sys/backtrace.rs:66:9
   3:     0x7e97171eab0d - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h148e7d8e87bab30f
                               at /rustc/2048386fe2898febe7315c0feb915458e41c7aa5/library/std/src/sys/backtrace.rs:39:26
   4:     0x7e971723b29b - core::fmt::rt::Argument::fmt::h7d54e75d5909de1b
                               at /rustc/2048386fe2898febe7315c0feb915458e41c7aa5/library/core/src/fmt/rt.rs:173:76
   5:     0x7e971723b29b - core::fmt::write::hf6482455a45a91c7
                               at /rustc/2048386fe2898febe7315c0feb915458e41c7aa5/library/core/src/fmt/mod.rs:1178:21
   6:     0x7e97171de9a3 - std::io::Write::write_fmt::h78337f43013967dc
                               at /rustc/2048386fe2898febe7315c0feb915458e41c7aa5/library/std/src/io/mod.rs:1823:15
   7:     0x7e97171ed302 - std::sys::backtrace::BacktraceLock::print::he76ad41bfebe12ae
                               at /rustc/2048386fe2898febe7315c0feb915458e41c7aa5/library/std/src/sys/backtrace.rs:42:9
   8:     0x7e97171ed302 - std::panicking::default_hook::{{closure}}::hea2fb4697dfb5a9e
                               at /rustc/2048386fe2898febe7315c0feb915458e41c7aa5/library/std/src/panicking.rs:266:22
   9:     0x7e97171ecf6e - std::panicking::default_hook::h1dd19741d2e30ca9
                               at /rustc/2048386fe2898febe7315c0feb915458e41c7aa5/library/std/src/panicking.rs:293:9
  10:     0x7e97136ba397 - std[3ddd7f8965915055]::panicking::update_hook::<alloc[5625efa5c1f04643]::boxed::Box<rustc_driver_impl[a075302d1f094aad]::install_ice_hook::{closure#0}>>::{closure#0}
  11:     0x7e97171edcf2 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h0b220dc3792e0071
                               at /rustc/2048386fe2898febe7315c0feb915458e41c7aa5/library/alloc/src/boxed.rs:2164:9
  12:     0x7e97171edcf2 - std::panicking::rust_panic_with_hook::h31e66c1e0f0e6e18
                               at /rustc/2048386fe2898febe7315c0feb915458e41c7aa5/library/std/src/panicking.rs:805:13
  13:     0x7e97171ed9a7 - std::panicking::begin_panic_handler::{{closure}}::h1057116e4a630d6a
                               at /rustc/2048386fe2898febe7315c0feb915458e41c7aa5/library/std/src/panicking.rs:671:13
  14:     0x7e97171eaff9 - std::sys::backtrace::__rust_end_short_backtrace::h943392b86d4267f1
                               at /rustc/2048386fe2898febe7315c0feb915458e41c7aa5/library/std/src/sys/backtrace.rs:170:18
  15:     0x7e97171ed634 - rust_begin_unwind
                               at /rustc/2048386fe2898febe7315c0feb915458e41c7aa5/library/std/src/panicking.rs:662:5
  16:     0x7e97172378f3 - core::panicking::panic_fmt::h8f072a14c223852d
                               at /rustc/2048386fe2898febe7315c0feb915458e41c7aa5/library/core/src/panicking.rs:74:14
  17:     0x7e9717237ae7 - core::panicking::panic_bounds_check::he6402fc016b56f3d
                               at /rustc/2048386fe2898febe7315c0feb915458e41c7aa5/library/core/src/panicking.rs:276:5
  18:     0x7e97139cf86a - <rustc_hir_typeck[dc91aa33eceb2355]::fn_ctxt::FnCtxt>::label_fn_like
  19:     0x7e97139c7356 - <rustc_hir_typeck[dc91aa33eceb2355]::fn_ctxt::FnCtxt>::report_arg_errors
  20:     0x7e97114bd534 - <rustc_hir_typeck[dc91aa33eceb2355]::fn_ctxt::FnCtxt>::check_method_argument_types
  21:     0x7e971550403c - <rustc_hir_typeck[dc91aa33eceb2355]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  22:     0x7e97154fba4b - <rustc_hir_typeck[dc91aa33eceb2355]::fn_ctxt::FnCtxt>::check_block_with_expected
  23:     0x7e97155032d8 - <rustc_hir_typeck[dc91aa33eceb2355]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  24:     0x7e97154de180 - rustc_hir_typeck[dc91aa33eceb2355]::check::check_fn
  25:     0x7e97154c08f7 - rustc_hir_typeck[dc91aa33eceb2355]::typeck
  26:     0x7e97154c02b3 - rustc_query_impl[d0e7c8c52ee208c5]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[d0e7c8c52ee208c5]::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[d6a66d085659a7e0]::query::erase::Erased<[u8; 8usize]>>
  27:     0x7e9714f633f5 - rustc_query_system[bcd77f19768ca8fb]::query::plumbing::try_execute_query::<rustc_query_impl[d0e7c8c52ee208c5]::DynamicConfig<rustc_query_system[bcd77f19768ca8fb]::query::caches::VecCache<rustc_span[d8ab92baa90519c6]::def_id::LocalDefId, rustc_middle[d6a66d085659a7e0]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[d0e7c8c52ee208c5]::plumbing::QueryCtxt, false>
  28:     0x7e9714f625e3 - rustc_query_impl[d0e7c8c52ee208c5]::query_impl::typeck::get_query_non_incr::__rust_end_short_backtrace
  29:     0x7e9714f62200 - <rustc_middle[d6a66d085659a7e0]::hir::map::Map>::par_body_owners::<rustc_hir_analysis[f363d87ce7e00737]::check_crate::{closure#4}>::{closure#0}
  30:     0x7e9714f5ffec - rustc_hir_analysis[f363d87ce7e00737]::check_crate
  31:     0x7e9714f55995 - rustc_interface[ecbe5e558f1e13b6]::passes::analysis
  32:     0x7e9714f55547 - rustc_query_impl[d0e7c8c52ee208c5]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[d0e7c8c52ee208c5]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[d6a66d085659a7e0]::query::erase::Erased<[u8; 1usize]>>
  33:     0x7e9715be502e - rustc_query_system[bcd77f19768ca8fb]::query::plumbing::try_execute_query::<rustc_query_impl[d0e7c8c52ee208c5]::DynamicConfig<rustc_query_system[bcd77f19768ca8fb]::query::caches::SingleCache<rustc_middle[d6a66d085659a7e0]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[d0e7c8c52ee208c5]::plumbing::QueryCtxt, false>
  34:     0x7e9715be4d8f - rustc_query_impl[d0e7c8c52ee208c5]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  35:     0x7e9715a5d260 - rustc_interface[ecbe5e558f1e13b6]::interface::run_compiler::<core[d8d378ef57b379bb]::result::Result<(), rustc_span[d8ab92baa90519c6]::ErrorGuaranteed>, rustc_driver_impl[a075302d1f094aad]::run_compiler::{closure#0}>::{closure#1}
  36:     0x7e971596a489 - std[3ddd7f8965915055]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[ecbe5e558f1e13b6]::util::run_in_thread_with_globals<rustc_interface[ecbe5e558f1e13b6]::util::run_in_thread_pool_with_globals<rustc_interface[ecbe5e558f1e13b6]::interface::run_compiler<core[d8d378ef57b379bb]::result::Result<(), rustc_span[d8ab92baa90519c6]::ErrorGuaranteed>, rustc_driver_impl[a075302d1f094aad]::run_compiler::{closure#0}>::{closure#1}, core[d8d378ef57b379bb]::result::Result<(), rustc_span[d8ab92baa90519c6]::ErrorGuaranteed>>::{closure#0}, core[d8d378ef57b379bb]::result::Result<(), rustc_span[d8ab92baa90519c6]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[d8d378ef57b379bb]::result::Result<(), rustc_span[d8ab92baa90519c6]::ErrorGuaranteed>>
  37:     0x7e971596a232 - <<std[3ddd7f8965915055]::thread::Builder>::spawn_unchecked_<rustc_interface[ecbe5e558f1e13b6]::util::run_in_thread_with_globals<rustc_interface[ecbe5e558f1e13b6]::util::run_in_thread_pool_with_globals<rustc_interface[ecbe5e558f1e13b6]::interface::run_compiler<core[d8d378ef57b379bb]::result::Result<(), rustc_span[d8ab92baa90519c6]::ErrorGuaranteed>, rustc_driver_impl[a075302d1f094aad]::run_compiler::{closure#0}>::{closure#1}, core[d8d378ef57b379bb]::result::Result<(), rustc_span[d8ab92baa90519c6]::ErrorGuaranteed>>::{closure#0}, core[d8d378ef57b379bb]::result::Result<(), rustc_span[d8ab92baa90519c6]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[d8d378ef57b379bb]::result::Result<(), rustc_span[d8ab92baa90519c6]::ErrorGuaranteed>>::{closure#1} as core[d8d378ef57b379bb]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  38:     0x7e97171f7a0b - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h8531124d63bf03d4
                               at /rustc/2048386fe2898febe7315c0feb915458e41c7aa5/library/alloc/src/boxed.rs:2150:9
  39:     0x7e97171f7a0b - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h14280c2e7cc937ab
                               at /rustc/2048386fe2898febe7315c0feb915458e41c7aa5/library/alloc/src/boxed.rs:2150:9
  40:     0x7e97171f7a0b - std::sys::pal::unix::thread::Thread::new::thread_start::hfbbba69bca48c3db
                               at /rustc/2048386fe2898febe7315c0feb915458e41c7aa5/library/std/src/sys/pal/unix/thread.rs:110:17
  41:     0x7e97100a6ded - <unknown>
  42:     0x7e971012a0dc - <unknown>
  43:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.82.0-nightly (2048386fe 2024-08-08) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `f`
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0425, E0601, E0658.
For more information about an error, try `rustc --explain E0425`.

```

</p>
</details>

<!--
query stack:
#0 [typeck] type-checking `f`
#1 [analysis] running analysis passes on this crate

-->


<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"jder"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_hir_typeck/src/callee.rs
+++ b/compiler/rustc_hir_typeck/src/callee.rs
@@ -58,7 +58,7 @@ pub(crate) fn check_legal_trait_for_method_call(
 enum CallStep<'tcx> {
     Builtin(Ty<'tcx>),
     DeferredClosure(LocalDefId, ty::FnSig<'tcx>),
-    /// E.g., enum variant constructors.
+    /// Call overloading when callee implements one of the Fn* traits.
     Overloaded(MethodCallee<'tcx>),
 }
 

--- a/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs
@@ -506,6 +506,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 fn_def_id,
                 call_span,
                 call_expr,
+                tuple_arguments,
             );
         }
     }
@@ -520,6 +521,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         fn_def_id: Option<DefId>,
         call_span: Span,
         call_expr: &'tcx hir::Expr<'tcx>,
+        tuple_arguments: TupleArgumentsFlag,
     ) -> ErrorGuaranteed {
         // Next, let's construct the error
         let (error_span, call_ident, full_call_span, call_name, is_method) = match &call_expr.kind {
@@ -865,6 +867,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         &matched_inputs,
                         &formal_and_expected_inputs,
                         is_method,
+                        tuple_arguments,
                     );
                     suggest_confusable(&mut err);
                     return err.emit();
@@ -1001,6 +1004,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 &matched_inputs,
                 &formal_and_expected_inputs,
                 is_method,
+                tuple_arguments,
             );
             suggest_confusable(&mut err);
             return err.emit();
@@ -1448,6 +1452,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             &matched_inputs,
             &formal_and_expected_inputs,
             is_method,
+            tuple_arguments,
         );
 
         // And add a suggestion block for all of the parameters
@@ -2219,21 +2224,27 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         matched_inputs: &IndexVec<ExpectedIdx, Option<ProvidedIdx>>,
         formal_and_expected_inputs: &IndexVec<ExpectedIdx, (Ty<'tcx>, Ty<'tcx>)>,
         is_method: bool,
+        tuple_arguments: TupleArgumentsFlag,
     ) {
         let Some(mut def_id) = callable_def_id else {
             return;
         };
 
+        // If we're calling a method of a Fn/FnMut/FnOnce trait object implicitly
+        // (eg invoking a closure) we want to point at the underlying callable,
+        // not the method implicitly invoked (eg call_once).
         if let Some(assoc_item) = self.tcx.opt_associated_item(def_id)
-            // Possibly points at either impl or trait item, so try to get it
-            // to point to trait item, then get the parent.
-            // This parent might be an impl in the case of an inherent function,
-            // but the next check will fail.
+            // Since this is an associated item, it might point at either an impl or a trait item.
+            // We want it to always point to the trait item.
+            // If we're pointing at an inherent function, we don't need to do anything,
+            // so we fetch the parent and verify if it's a trait item.
             && let maybe_trait_item_def_id = assoc_item.trait_item_def_id.unwrap_or(def_id)
             && let maybe_trait_def_id = self.tcx.parent(maybe_trait_item_def_id)
             // Just an easy way to check "trait_def_id == Fn/FnMut/FnOnce"
             && let Some(call_kind) = self.tcx.fn_trait_kind_from_def_id(maybe_trait_def_id)
             && let Some(callee_ty) = callee_ty
+            // TupleArguments is set only when this is an implicit call (my_closure(...)) rather than explicit (my_closure.call(...))
+            && tuple_arguments == TupleArguments
         {
             let callee_ty = callee_ty.peel_refs();
             match *callee_ty.kind() {
@@ -2303,81 +2314,154 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         {
             let mut spans: MultiSpan = def_span.into();
 
-            let params_with_generics = self.get_hir_params_with_generics(def_id, is_method);
-            let mut generics_with_unmatched_params = Vec::new();
+            if let Some(params_with_generics) = self.get_hir_params_with_generics(def_id, is_method)
+            {
+                debug_assert_eq!(params_with_generics.len(), matched_inputs.len());
 
-            let check_for_matched_generics = || {
-                if matched_inputs.iter().any(|x| x.is_some())
-                    && params_with_generics.iter().any(|x| x.0.is_some())
-                {
-                    for (idx, (generic, _)) in params_with_generics.iter().enumerate() {
-                        // Param has to have a generic and be matched to be relevant
-                        if matched_inputs[idx.into()].is_none() {
-                            continue;
-                        }
+                let mut generics_with_unmatched_params = Vec::new();
 
-                        let Some(generic) = generic else {
-                            continue;
-                        };
+                let check_for_matched_generics = || {
+                    if matched_inputs.iter().any(|x| x.is_some())
+                        && params_with_generics.iter().any(|x| x.0.is_some())
+                    {
+                        for (idx, (generic, _)) in params_with_generics.iter().enumerate() {
+                            // Param has to have a generic and be matched to be relevant
+                            if matched_inputs[idx.into()].is_none() {
+                                continue;
+                            }
 
-                        for unmatching_idx in idx + 1..params_with_generics.len() {
-                            if matched_inputs[unmatching_idx.into()].is_none()
-                                && let Some(unmatched_idx_param_generic) =
-                                    params_with_generics[unmatching_idx].0
-                                && unmatched_idx_param_generic.name.ident() == generic.name.ident()
-                            {
-                                // We found a parameter that didn't match that needed to
-                                return true;
+                            let Some(generic) = generic else {
+                                continue;
+                            };
+
+                            for unmatching_idx in idx + 1..params_with_generics.len() {
+                                if matched_inputs[unmatching_idx.into()].is_none()
+                                    && let Some(unmatched_idx_param_generic) =
+                                        params_with_generics[unmatching_idx].0
+                                    && unmatched_idx_param_generic.name.ident()
+                                        == generic.name.ident()
+                                {
+                                    // We found a parameter that didn't match that needed to
+                                    return true;
+                                }
                             }
                         }
                     }
-                }
-                false
-            };
-
-            let check_for_matched_generics = check_for_matched_generics();
-
-            for (idx, (generic_param, param)) in
-                params_with_generics.iter().enumerate().filter(|(idx, _)| {
-                    check_for_matched_generics
-                        || expected_idx.is_none_or(|expected_idx| expected_idx == *idx)
-                })
-            {
-                let Some(generic_param) = generic_param else {
-                    spans.push_span_label(param.span, "");
-                    continue;
+                    false
                 };
 
-                let other_params_matched: Vec<(usize, &hir::Param<'_>)> = params_with_generics
-                    .iter()
-                    .enumerate()
-                    .filter(|(other_idx, (other_generic_param, _))| {
-                        if *other_idx == idx {
-                            return false;
-                        }
-                        let Some(other_generic_param) = other_generic_param else {
-                            return false;
-                        };
-                        if matched_inputs[idx.into()].is_none()
-                            && matched_inputs[(*other_idx).into()].is_none()
-                        {
-                            return false;
-                        }
-                        if matched_inputs[idx.into()].is_some()
-                            && matched_inputs[(*other_idx).into()].is_some()
-                        {
-                            return false;
-                        }
-                        other_generic_param.name.ident() == generic_param.name.ident()
-                    })
-                    .map(|(other_idx, (_, other_param))| (other_idx, *other_param))
-                    .collect();
+                let check_for_matched_generics = check_for_matched_generics();
 
-                if !other_params_matched.is_empty() {
-                    let other_param_matched_names: Vec<String> = other_params_matched
+                for (idx, (generic_param, param)) in
+                    params_with_generics.iter().enumerate().filter(|(idx, _)| {
+                        check_for_matched_generics
+                            || expected_idx.is_none_or(|expected_idx| expected_idx == *idx)
+                    })
+                {
+                    let Some(generic_param) = generic_param else {
+                        spans.push_span_label(param.span, "");
+                        continue;
+                    };
+
+                    let other_params_matched: Vec<(usize, &hir::Param<'_>)> = params_with_generics
                         .iter()
-                        .map(|(_, other_param)| {
-                            if let hir::PatKind::Binding(_, _, ident, _) = other_param.pat.kind {
+                        .enumerate()
+                        .filter(|(other_idx, (other_generic_param, _))| {
+                            if *other_idx == idx {
+                                return false;
+                            }
+                            let Some(other_generic_param) = other_generic_param else {
+                                return false;
+                            };
+                            if matched_inputs[idx.into()].is_none()
+                                && matched_inputs[(*other_idx).into()].is_none()
+                            {
+                                return false;
+                            }
+                            if matched_inputs[idx.into()].is_some()
+                                && matched_inputs[(*other_idx).into()].is_some()
+                            {
+                                return false;
+                            }
+                            other_generic_param.name.ident() == generic_param.name.ident()
+                        })
+                        .map(|(other_idx, (_, other_param))| (other_idx, *other_param))
+                        .collect();
+
+                    if !other_params_matched.is_empty() {
+                        let other_param_matched_names: Vec<String> = other_params_matched
+                            .iter()
+                            .map(|(_, other_param)| {
+                                if let hir::PatKind::Binding(_, _, ident, _) = other_param.pat.kind
+                                {
+                                    format!("`{ident}`")
+                                } else {
+                                    "{unknown}".to_string()
+                                }
+                            })
+                            .collect();
+
+                        let matched_ty = self
+                            .resolve_vars_if_possible(formal_and_expected_inputs[idx.into()].1)
+                            .sort_string(self.tcx);
+
+                        if matched_inputs[idx.into()].is_some() {
+                            spans.push_span_label(
+                                param.span,
+                                format!(
+                                    "{} {} to match the {} type of this parameter",
+                                    display_list_with_comma_and(&other_param_matched_names),
+                                    format!(
+                                        "need{}",
+                                        pluralize!(if other_param_matched_names.len() == 1 {
+                                            0
+                                        } else {
+                                            1
+                                        })
+                                    ),
+                                    matched_ty,
+                                ),
+                            );
+                        } else {
+                            spans.push_span_label(
+                                param.span,
+                                format!(
+                                    "this parameter needs to match the {} type of {}",
+                                    matched_ty,
+                                    display_list_with_comma_and(&other_param_matched_names),
+                                ),
+                            );
+                        }
+                        generics_with_unmatched_params.push(generic_param);
+                    } else {
+                        spans.push_span_label(param.span, "");
+                    }
+                }
+
+                for generic_param in self
+                    .tcx
+                    .hir()
+                    .get_if_local(def_id)
+                    .and_then(|node| node.generics())
+                    .into_iter()
+                    .flat_map(|x| x.params)
+                    .filter(|x| {
+                        generics_with_unmatched_params
+                            .iter()
+                            .any(|y| x.name.ident() == y.name.ident())
+                    })
+                {
+                    let param_idents_matching: Vec<String> = params_with_generics
+                        .iter()
+                        .filter(|(generic, _)| {
+                            if let Some(generic) = generic {
+                                generic.name.ident() == generic_param.name.ident()
+                            } else {
+                                false
+                            }
+                        })
+                        .map(|(_, param)| {
+                            if let hir::PatKind::Binding(_, _, ident, _) = param.pat.kind {
                                 format!("`{ident}`")
                             } else {
                                 "{unknown}".to_string()
@@ -2385,84 +2469,18 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         })
                         .collect();
 
-                    let matched_ty = self
-                        .resolve_vars_if_possible(formal_and_expected_inputs[idx.into()].1)
-                        .sort_string(self.tcx);
-
-                    if matched_inputs[idx.into()].is_some() {
+                    if !param_idents_matching.is_empty() {
                         spans.push_span_label(
-                            param.span,
+                            generic_param.span,
                             format!(
-                                "{} {} to match the {} type of this parameter",
-                                display_list_with_comma_and(&other_param_matched_names),
-                                format!(
-                                    "need{}",
-                                    pluralize!(if other_param_matched_names.len() == 1 {
-                                        0
-                                    } else {
-                                        1
-                                    })
-                                ),
-                                matched_ty,
-                            ),
-                        );
-                    } else {
-                        spans.push_span_label(
-                            param.span,
-                            format!(
-                                "this parameter needs to match the {} type of {}",
-                                matched_ty,
-                                display_list_with_comma_and(&other_param_matched_names),
+                                "{} all reference this parameter {}",
+                                display_list_with_comma_and(&param_idents_matching),
+                                generic_param.name.ident().name,
                             ),
                         );
                     }
-                    generics_with_unmatched_params.push(generic_param);
-                } else {
-                    spans.push_span_label(param.span, "");
                 }
             }
-
-            for generic_param in self
-                .tcx
-                .hir()
-                .get_if_local(def_id)
-                .and_then(|node| node.generics())
-                .into_iter()
-                .flat_map(|x| x.params)
-                .filter(|x| {
-                    generics_with_unmatched_params.iter().any(|y| x.name.ident() == y.name.ident())
-                })
-            {
-                let param_idents_matching: Vec<String> = params_with_generics
-                    .iter()
-                    .filter(|(generic, _)| {
-                        if let Some(generic) = generic {
-                            generic.name.ident() == generic_param.name.ident()
-                        } else {
-                            false
-                        }
-                    })
-                    .map(|(_, param)| {
-                        if let hir::PatKind::Binding(_, _, ident, _) = param.pat.kind {
-                            format!("`{ident}`")
-                        } else {
-                            "{unknown}".to_string()
-                        }
-                    })
-                    .collect();
-
-                if !param_idents_matching.is_empty() {
-                    spans.push_span_label(
-                        generic_param.span,
-                        format!(
-                            "{} all reference this parameter {}",
-                            display_list_with_comma_and(&param_idents_matching),
-                            generic_param.name.ident().name,
-                        ),
-                    );
-                }
-            }
-
             err.span_note(spans, format!("{} defined here", self.tcx.def_descr(def_id)));
         } else if let Some(hir::Node::Expr(e)) = self.tcx.hir().get_if_local(def_id)
             && let hir::ExprKind::Closure(hir::Closure { body, .. }) = &e.kind
@@ -2535,74 +2553,77 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             return;
         };
 
-        let params_with_generics = self.get_hir_params_with_generics(def_id, is_method);
+        if let Some(params_with_generics) = self.get_hir_params_with_generics(def_id, is_method) {
+            debug_assert_eq!(params_with_generics.len(), matched_inputs.len());
+            for (idx, (generic_param, _)) in params_with_generics.iter().enumerate() {
+                if matched_inputs[idx.into()].is_none() {
+                    continue;
+                }
 
-        for (idx, (generic_param, _)) in params_with_generics.iter().enumerate() {
-            if matched_inputs[idx.into()].is_none() {
-                continue;
+                let Some((_, matched_arg_span)) = provided_arg_tys.get(idx.into()) else {
+                    continue;
+                };
+
+                let Some(generic_param) = generic_param else {
+                    continue;
+                };
+
+                let mut idxs_matched: Vec<usize> = vec![];
+                for (other_idx, (_, _)) in params_with_generics.iter().enumerate().filter(
+                    |(other_idx, (other_generic_param, _))| {
+                        if *other_idx == idx {
+                            return false;
+                        }
+                        let Some(other_generic_param) = other_generic_param else {
+                            return false;
+                        };
+                        if matched_inputs[(*other_idx).into()].is_some() {
+                            return false;
+                        }
+                        other_generic_param.name.ident() == generic_param.name.ident()
+                    },
+                ) {
+                    idxs_matched.push(other_idx);
+                }
+
+                if idxs_matched.is_empty() {
+                    continue;
+                }
+
+                let expected_display_type = self
+                    .resolve_vars_if_possible(formal_and_expected_inputs[idx.into()].1)
+                    .sort_string(self.tcx);
+                let label = if idxs_matched.len() == params_with_generics.len() - 1 {
+                    format!(
+                        "expected all arguments to be this {} type because they need to match the type of this parameter",
+                        expected_display_type
+                    )
+                } else {
+                    format!(
+                        "expected some other arguments to be {} {} type to match the type of this parameter",
+                        a_or_an(&expected_display_type),
+                        expected_display_type,
+                    )
+                };
+
+                err.span_label(*matched_arg_span, label);
             }
-
-            let Some((_, matched_arg_span)) = provided_arg_tys.get(idx.into()) else {
-                continue;
-            };
-
-            let Some(generic_param) = generic_param else {
-                continue;
-            };
-
-            let mut idxs_matched: Vec<usize> = vec![];
-            for (other_idx, (_, _)) in params_with_generics.iter().enumerate().filter(
-                |(other_idx, (other_generic_param, _))| {
-                    if *other_idx == idx {
-                        return false;
-                    }
-                    let Some(other_generic_param) = other_generic_param else {
-                        return false;
-                    };
-                    if matched_inputs[(*other_idx).into()].is_some() {
-                        return false;
-                    }
-                    other_generic_param.name.ident() == generic_param.name.ident()
-                },
-            ) {
-                idxs_matched.push(other_idx);
-            }
-
-            if idxs_matched.is_empty() {
-                continue;
-            }
-
-            let expected_display_type = self
-                .resolve_vars_if_possible(formal_and_expected_inputs[idx.into()].1)
-                .sort_string(self.tcx);
-            let label = if idxs_matched.len() == params_with_generics.len() - 1 {
-                format!(
-                    "expected all arguments to be this {} type because they need to match the type of this parameter",
-                    expected_display_type
-                )
-            } else {
-                format!(
-                    "expected some other arguments to be {} {} type to match the type of this parameter",
-                    a_or_an(&expected_display_type),
-                    expected_display_type,
-                )
-            };
-
-            err.span_label(*matched_arg_span, label);
         }
     }
 
+    /// Returns the parameters of a function, with their generic parameters if those are the full
+    /// type of that parameter. Returns `None` if the function body is unavailable (eg is an instrinsic).
     fn get_hir_params_with_generics(
         &self,
         def_id: DefId,
         is_method: bool,
-    ) -> Vec<(Option<&hir::GenericParam<'_>>, &hir::Param<'_>)> {
-        let fn_node = self.tcx.hir().get_if_local(def_id);
+    ) -> Option<Vec<(Option<&hir::GenericParam<'_>>, &hir::Param<'_>)>> {
+        let fn_node = self.tcx.hir().get_if_local(def_id)?;
 
         let generic_params: Vec<Option<&hir::GenericParam<'_>>> = fn_node
-            .and_then(|node| node.fn_decl())
+            .fn_decl()?
+            .inputs
             .into_iter()
-            .flat_map(|decl| decl.inputs)
             .skip(if is_method { 1 } else { 0 })
             .map(|param| {
                 if let hir::TyKind::Path(QPath::Resolved(
@@ -2611,7 +2632,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 )) = param.kind
                 {
                     fn_node
-                        .and_then(|node| node.generics())
+                        .generics()
                         .into_iter()
                         .flat_map(|generics| generics.params)
                         .find(|param| &param.def_id.to_def_id() == res_def_id)
@@ -2621,14 +2642,16 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             })
             .collect();
 
-        let params: Vec<&hir::Param<'_>> = fn_node
-            .and_then(|node| node.body_id())
+        let params: Vec<&hir::Param<'_>> = self
+            .tcx
+            .hir()
+            .body(fn_node.body_id()?)
+            .params
             .into_iter()
-            .flat_map(|id| self.tcx.hir().body(id).params)
             .skip(if is_method { 1 } else { 0 })
             .collect();
 
-        generic_params.into_iter().zip(params).collect()
+        Some(generic_params.into_iter().zip_eq(params).collect())
     }
 }
 

--- a/tests/crashes/128848.rs
+++ b/tests/crashes/128848.rs
@@ -1,5 +1,0 @@
-//@ known-bug: rust-lang/rust#128848
-
-fn f<T>(a: T, b: T, c: T)  {
-    f.call_once()
-}

--- a/tests/ui/mismatched_types/mismatch-args-crash-issue-128848.rs
+++ b/tests/ui/mismatched_types/mismatch-args-crash-issue-128848.rs
@@ -1,0 +1,10 @@
+#![feature(fn_traits)]
+
+// Regression test for https://github.com/rust-lang/rust/issues/128848
+
+fn f<T>(a: T, b: T, c: T) {
+    f.call_once()
+    //~^ ERROR this method takes 1 argument but 0 arguments were supplied
+}
+
+fn main() {}

--- a/tests/ui/mismatched_types/mismatch-args-crash-issue-128848.stderr
+++ b/tests/ui/mismatched_types/mismatch-args-crash-issue-128848.stderr
@@ -1,0 +1,16 @@
+error[E0061]: this method takes 1 argument but 0 arguments were supplied
+  --> $DIR/mismatch-args-crash-issue-128848.rs:6:7
+   |
+LL |     f.call_once()
+   |       ^^^^^^^^^-- argument #1 of type `(_, _, _)` is missing
+   |
+note: method defined here
+  --> $SRC_DIR/core/src/ops/function.rs:LL:COL
+help: provide the argument
+   |
+LL |     f.call_once(/* args */)
+   |                ~~~~~~~~~~~~
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0061`.

